apluslms_roman/backends/__init__.py

```diff
@@ -67,21 +67,25 @@ def __str__(self):
 
 
 class BuildResult:
-    __slots__ = ('code', 'error', 'step')
+    __slots__ = ('step', 'ok', 'code', 'error')
 
-    def __init__(self, code=0, error=None, step=None):
-        self.code = code
-        self.error = error
-        self.step = step
-        assert self.ok or step is not None, "step is required for failed result"
+    def __init__(self, step, ok=True, code=0, error=None):
+        self.step, self.ok, self.code, self.error = (
+            step, ok, code, error)
 
     @property
-    def ok(self):
-        return self.code == 0 and self.error is None
+    def cancelled(self):
+        return not self.ok and self.code == 0
+
+    @property
+    def failed(self):
+        return not self.ok and self.code != 0
 
     def __str__(self):
         if self.ok:
             return "Build ok"
+        if self.cancelled:
+            return "Build cancelled"
```

Comment:
Cancelled and failed both have step they are related to, thus the default print could be same (with different stateus) for both of them. On the other hand, ok result doesn't have step, so it can't be mentioned.

Reply:
Do you mean that the str for when a step is cancelled shoulb be "Build failed on step x"? Or "Build cancelled on step x"?

Reply:
Second

Reply:
should the observer's output have the same format? if yes, it would probably make more sense to directly use the BuildResult's str in the output

Reply:
If I remember right, the observer loses this class before printing, so it's hard to use the same info. This str formatting probably isn't even used by cli. Possibly by GUI.

Reply:
It used to be used in cli (`print(result)`), but isn't anymore. The `result_msg` function could be changed to take a BuildResult as input, if that doesn't complicate things too much.

You didn't really answer wether or not you wanted me to change the output when build is cancelled in observer, too?

Reply:
> The result_msg function could be changed to take a BuildResult as input,

It does, but it breaks it to base types, so it could be serialized (keeps the _message protocol simple).

> wether or not you wanted me to change the output when build is cancelled in observer, too?

I think the streamobserver output is pretty much ok. It could write similar conclusion line for cancelled case too.